src/osd/scrubber/scrub_backend.h

```diff
@@ -0,0 +1,419 @@
+  }
+};
+
+/**
+ * the back-end data that is per-chunk
+ *
+ * Created by the Scrubber after all replicas' maps have arrived.
+ */
+struct ScrubBeChunk {
+
+  explicit ScrubBeChunk(pg_shard_t i_am)
+  {
+    received_maps[i_am] = ScrubMap{};
+  }
+
+  /// the working set of scrub maps: the received maps, plus
+  /// Primary's own map.
+  std::map<pg_shard_t, ScrubMap> received_maps;
+
+  /// a collection of all objs mentioned in the maps
+  std::set<hobject_t> master_set;
+
+  utime_t started{ceph_clock_now()};
+
+  std::map<hobject_t, data_omap_digests_t> missing_digest;
+
+  /// Map from object with errors to good peers
+  std::map<hobject_t, std::list<pg_shard_t>> authoritative;
+
+
+  // these must be reset for each element:
+
+  std::set<pg_shard_t> cur_missing;
+  std::set<pg_shard_t> cur_inconsistent;
+  bool fix_digest{false};
+};
+
+
+/**
+ * ScrubBackend wraps the data and operations required for the back-end part of
+ * the scrubbing (i.e. for comparing the maps and fixing objects).
+ *
+ * Created anew upon each initiation of a scrub session.
+ *
+ * Uses 'optional' entries for winking-out the auxiliary collections.
+ */
+class ScrubBackend : public ScrubBackendIF {
+ public:
+  // Primary constructor
+  ScrubBackend(PgScrubber& scrubber,
+               PGBackend& backend,
+               PG& pg,
+               pg_shard_t i_am,
+               bool repair,
+               scrub_level_t shallow_or_deep,
+               const std::set<pg_shard_t>& acting);
+
+  // Replica constructor: no primary map
+  ScrubBackend(PgScrubber& scrubber,
+               PGBackend& backend,
+               PG& pg,
+               pg_shard_t i_am,
+               bool repair,
+               scrub_level_t shallow_or_deep);
+
+  friend class PgScrubber;
+
+  /**
+   * reset the per-chunk data structure (ScrubberBeChunk).
+   * Create an empty scrub-map for this shard, and place it
+   * in the appropriate entry in 'received_maps'.
+   *
+   * @returns a pointer to the newly created ScrubMap.
+   */
+  ScrubMap* new_chunk() final;
+
+  /**
+   * sets Backend's m_repair flag (setting m_mode_desc to a corresponding
+   * string)
+   */
+  void update_repair_status(bool should_repair) final;
+
+  // note: used by both Primary & replicas
+  static ScrubMap clean_meta_map(ScrubMap& cleaned, bool max_reached);
+
+  void replica_clean_meta(ScrubMap& smap,
+                          bool max_reached,
+                          const hobject_t& start) final;
+
+  /**
+   * decode the arriving MOSDRepScrubMap message, placing the replica's
+   * scrub-map into received_maps[from].
+   *
+   * @param from replica
+   * @param pool the pool the replica is in
+   */
+  void decode_received_map(pg_shard_t from, const MOSDRepScrubMap& msg) final;
+
+  void scrub_compare_maps(bool max_reached) final;
+
+  int get_num_digest_updates_pending() const final
+  {
+    return num_digest_updates_pending;
+  }
+
+  void scan_snaps(ScrubMap& smap) final;
```

Comment:
Not used externally, shouldn't be public.

Reply:
fixed